Automate EC2 based staging environment for StrataSTOR

yendelevium opened the pull request:
- Create an EC2 based staging environment via terraform
- Create a VPC, internet gateway, route table, subnet, security group, network interface, elastic IP and an EC2
- Configure the EC2 via the user_data shell script
- Add a README documenting deployment, validation, and destruction (in forge/infra)
- Add a .gitgnore file (sourced from a github template)

It was assumed that there are existing VPC(s), subnets, NAT, and Internet Gateways (if private subnet is used), but I wanted the staging environment to have it's own configurable(and automated) VPC, subnet etc, so I'm creating them in terraform from scratch.

Resolves #1

yendelevium committed:
Automate EC2 based staging environment for StrataSTOR

- Create an EC2 based staging environment via terraform
- Create a VPC, internet gateway, route table, subnet, security group, network interface, elastic IP and an EC2
- Configure the EC2 via the user_data shell script
- Add a README documenting deployment, validation, and destruction (in forge/infra)
- Add a .gitgnore file (sourced from a github template)

Resolves #1

yendelevium commented:
@raamsri  I have created the staging environment, but I'm concerned about the following
- Terraform code style : I'm not sure how to write properly structured terraform code
- Proper permissions : I have very basic knowledge about the cloud, and most of my code is taken from the terraform docs(like the security groups and internet gateways etc), so I couldn't fine-tune the environment. But passes all the tests, and everything required is installed
- The build take upwards of 25 mins to complete, most of the time going towards downloading zfs 2.3.0-rc5. Again, I'm not sure if it's supposed to take this long.

Please review my code and test it out yourself, I believe it does the basic tasks but will need a lot more fine-tuning

yendelevium commented:
Also, for some reason it seems to have modified the code-of-conduct, contributing and some other docs, but they are the same as in the main branch, please ignore them.

raamsri commented:
I appreciate the effort and I understand that you are just getting acquainted with AWS and its ecosystem 🙂 

> It was assumed that there are existing VPC(s), subnets, NAT, and Internet Gateways (if private subnet is used), but I wanted the staging environment to have it's own configurable(and automated) VPC, subnet etc, so I'm creating them in terraform from scratch.
Lack of flexibility to play well with an existing setup limits the scope of development. Testing may be carried out on multiple machines, and it may require interaction with other services that are deployed somewhere. Creating new VPCs, and the other necessary resources along with it, for every staging instance is not convenient. Besides that, lack of configuration flexibility may also prove to be expensive apart from staging inconvenience.

raamsri commented:
I appreciate the effort and I understand that you are just getting acquainted with AWS and its ecosystem 🙂 

> It was assumed that there are existing VPC(s), subnets, NAT, and Internet Gateways (if private subnet is used), but I wanted the staging environment to have it's own configurable(and automated) VPC, subnet etc, so I'm creating them in terraform from scratch.

Lack of flexibility to play well with an existing setup limits the scope of development. Testing may be carried out on multiple machines, and it may require interaction with other services that are deployed somewhere. Creating new VPCs, and the other necessary resources along with it, for every staging instance is not convenient. Besides that, lack of configuration flexibility may also prove to be expensive apart from staging inconvenience.

raamsri commented:
The idea is to accommodate a convenient development pipeline. Perhaps continue using the scripts while developing Rodent to get a sense for its use and application? This primitive setup is likely to improve with continuous development in a useful way.

raamsri commented:
> Perhaps continue using the scripts while developing Rodent to get a sense for its use and application?

This is how my dev workflow happens at the moment: I have a staging ec2 for rodent that's in a VPC along with other projects. I write code on my Mac, `rsync` that code to the ec2 and run the service and tests in it. VSCode allows remote development but I'm not using it so as to spare the server's compute and memory for other needs. I operate the machine,  primarily to turn on/off, execute commands over ssh and sync code, via custom shell scripts/functions for convenience. I spin-up another Rodent machine when I test data transfer modules. More ad-hoc machines may be needed as required for testing.

yendelevium commented:
> > Perhaps continue using the scripts while developing Rodent to get a sense for its use and application?
> 
> This is how my dev workflow happens at the moment: I have a staging ec2 for rodent that's in a VPC along with other projects. I write code on my Mac, `rsync` that code to the ec2 and run the service and tests in it. VSCode allows remote development but I'm not using it so as to spare the server's compute and memory for other needs. I operate the machine, primarily to turn on/off, execute commands over ssh and sync code, via custom shell scripts/functions for convenience. I spin-up another Rodent machine when I test data transfer modules. More ad-hoc machines may be needed as required for testing.

I understand

yendelevium commented:
> I appreciate the effort and I understand that you are just getting acquainted with AWS and its ecosystem 🙂
> 
> > It was assumed that there are existing VPC(s), subnets, NAT, and Internet Gateways (if private subnet is used), but I wanted the staging environment to have it's own configurable(and automated) VPC, subnet etc, so I'm creating them in terraform from scratch.
> 
> Lack of flexibility to play well with an existing setup limits the scope of development. Testing may be carried out on multiple machines, and it may require interaction with other services that are deployed somewhere. Creating new VPCs, and the other necessary resources along with it, for every staging instance is not convenient. Besides that, lack of configuration flexibility may also prove to be expensive apart from staging inconvenience.

Thank you for clearing this up.
So i guess we just need to automate the creating of an EC2, and we associate a network interface with it, and then run the scripts?

yendelevium commented:
Should I proceed @raamsri ?

yendelevium committed:
Remove VPC, Subnet, IG creation

- Remove the code for creating VPC, subnets, NAT and IGs
- Add dependency for network interface as variable
- Update README

yendelevium commented:
Please review @raamsri

raamsri commented:
Thanks for the effort @yendelevium . There's plenty of room for improvement. Unfortunately, explaining the required, and the expected, quality of wiring to meet modular IaC standards would be redundant as there are plenty of HowTos out there already. And, I would appreciate a clean commit that doesn't upset files which aren't in context; I'm closing this PR.

yendelevium commented:
I understand, I'll try again with better understanding next time. Thank you for the feedback
